rclcpp_components/src/component_container_mt.cpp

```diff
@@ -22,8 +22,19 @@ int main(int argc, char * argv[])
 {
   /// Component container with a multi-threaded executor.
   rclcpp::init(argc, argv);
+
+  auto options = rclcpp::NodeOptions{};
   auto exec = std::make_shared<rclcpp::executors::MultiThreadedExecutor>();
-  auto node = std::make_shared<rclcpp_components::ComponentManager>(exec);
+  auto node = std::make_shared<rclcpp_components::ComponentManager>(
+    std::weak_ptr<rclcpp::executors::MultiThreadedExecutor>(), "ComponentManager", options);
```

Comment:
I think that we can delete the `std::weak_ptr<rclcpp::executors::MultiThreadedExecutor>` argument from the component manager constructor, and initialize the internal member to `nullptr` there.

Reply:
> initialize the internal member to nullptr there.

This causes a build error.

Reply:
Can you copy the build error?

Initializing a weak ptr to nullptr is valid.
See https://en.cppreference.com/w/cpp/memory/weak_ptr/weak_ptr (overload (1)).

Reply:
```
/home/daisuke/workspace/sandbox_ws/src/rclcpp/rclcpp_components/include/rclcpp_components/component_manager.hpp:102:48: err│
or: could not convert ‘nullptr’ from ‘std::nullptr_t’ to ‘std::weak_ptr<rclcpp::Executor>’                                 │
  102 |     std::weak_ptr<rclcpp::Executor> executor = nullptr,                                                            │
      |                                                ^~~~~~~                                                             │
      |                                                |                                                                   │
      |                                                std::nullptr_t                                                      │
make[2]: *** [CMakeFiles/component_manager.dir/build.make:63: CMakeFiles/component_manager.dir/src/component_manager.cpp.o]│
 Error 1                                                                                                                   │
make[1]: *** [CMakeFiles/Makefile2:138: CMakeFiles/component_manager.dir/all] Error 2                                      │
make[1]: *** Waiting for unfinished jobs....                                                                               │
make: *** [Makefile:141: all] Error 2                                                                                      │
```

Reply:
```
││   ComponentManager(   ComponentManager(   ComponentManager(                                                               ││
│y                                                                                  │││-    std::weak_ptr<rclcpp::Executor> executor,                                                                             ││
│y                                                                                  │││+    std::weak_ptr<rclcpp::Executor> executor = nullptr,                                                                   ││
│y                                                                                  │││     std::string node_name = "ComponentManager",                                                                           ││
│y                                                                                  │││     const rclcpp::NodeOptions & node_options = rclcpp::NodeOptions()                                                      ││
│y                                                                                  │││     .start_parameter_services(false)                                                                                      ││
│y                                                                                  │││diff --git a/rclcpp_components/src/component_container_mt.cpp b/rclcpp_components/src/component_container_mt.cpp           ││
│y                                                                                  │││index 8c32c08f..013c0b1c 100644                                                                                            ││
│y                                                                                  │││--- a/rclcpp_components/src/component_container_mt.cpp                                                                     ││
│y                                                                                  │││+++ b/rclcpp_components/src/component_container_mt.cpp                                                                     ││
│y                                                                                  │││@@ -25,8 +25,7 @@ int main(int argc, char * argv[])                                                                        ││
│y                                                                                  │││                                                                                                                           ││
│y                                                                                  │││   auto options = rclcpp::NodeOptions{};                                                                                   ││
│y                                                                                  │││   auto exec = std::make_shared<rclcpp::executors::MultiThreadedExecutor>();                                               ││
│y                                                                                  │││-  auto node = std::make_shared<rclcpp_components::ComponentManager>(                                                      ││
│y                                                                                  │││-    std::weak_ptr<rclcpp::executors::MultiThreadedExecutor>(), "ComponentManager", options);                              ││
│y                                                                                  │││+  auto node = std::make_shared<rclcpp_components::ComponentManager>();                                                    ││
│y                                                                                  │││   if (node->has_parameter("thread_num")) {                                                                                ││
│y                                                                                  │││     const auto thread_num = node->get_parameter("thread_num").as_int();                                                   ││
│y                                                                                  │││     exec = std::make_shared<rclcpp::executors::MultiThreadedExecutor>( 
```

Reply:
```
~/workspace/sandbox_ws/src/rclcpp feature/add-param-for-thread-num* 9s                                                     ││                                                                                                                           │
│❯ git diff                                                                                                                 ││                                                                                                                           │
│diff --git a/rclcpp_components/include/rclcpp_components/component_manager.hpp b/rclcpp_components/include/rclcpp_component││                                                                                                                           │
│s/component_manager.hpp                                                                                                    ││                                                                                                                           │
│index 67dc90c4..3d9c7559 100644                                                                                            ││                                                                                                                           │
│--- a/rclcpp_components/include/rclcpp_components/component_manager.hpp                                                    ││                                                                                                                           │
│+++ b/rclcpp_components/include/rclcpp_components/component_manager.hpp                                                    ││                                                                                                                           │
│@@ -99,7 +99,7 @@ public:                                                                                                  ││                                                                                                                           │
│    */                                                                                                                     ││                                                                                                                           │
│   RCLCPP_COMPONENTS_PUBLIC                                                                                                ││                                                                                                                           │
│   ComponentManager(                                                                                                       ││                                                                                                                           │
│-    std::weak_ptr<rclcpp::Executor> executor,                                                                             ││                                                                                                                           │
│+    std::weak_ptr<rclcpp::Executor> executor = nullptr,                                                                   ││                                                                                                                           │
│     std::string node_name = "ComponentManager",                                                                           ││                                                                                                                           │
│     const rclcpp::NodeOptions & node_options = rclcpp::NodeOptions()                                                      ││                                                                                                                           │
│     .start_parameter_services(false)                                                                                      ││                                                                                                                           │
│diff --git a/rclcpp_components/src/component_container_mt.cpp b/rclcpp_components/src/component_container_mt.cpp           ││                                                                                                                           │
│index 8c32c08f..013c0b1c 100644                                                                                            ││                                                                                                                           │
│--- a/rclcpp_components/src/component_container_mt.cpp                                                                     ││                                                                                                                           │
│+++ b/rclcpp_components/src/component_container_mt.cpp                                                                     ││                                                                                                                           │
│@@ -25,8 +25,7 @@ int main(int argc, char * argv[])                                                                        ││                                                                                                                           │
│                                                                                                                           ││                                                                                                                           │
│   auto options = rclcpp::NodeOptions{};         
```

Reply:
> Initializing a weak ptr to nullptr is valid.
See https://en.cppreference.com/w/cpp/memory/weak_ptr/weak_ptr (overload (1)).

I believe that default constructs an empty `weak_ptr` but that isn't a `nullptr` - just an empty `weak_ptr<rclcpp::Executor>`.

Shouldn't a `weak_ptr` necessarily only be initialized from another `weak_ptr` or `shared_ptr`? Initializing with `nullptr` (or a raw pointer for that matter) would defeat the point.

Reply:
```
~/workspace/sandbox_ws/src/rclcpp feature/add-param-for-thread-num* 9s                                                     ││                                                                                                                           │
│❯ git diff                                                                                                                 ││                                                                                                                           │
│diff --git a/rclcpp_components/include/rclcpp_components/component_manager.hpp b/rclcpp_components/include/rclcpp_component││                                                                                                                           │
│s/component_manager.hpp                                                                                                    ││                                                                                                                           │
│index 67dc90c4..3d9c7559 100644                                                                                            ││                                                                                                                           │
│--- a/rclcpp_components/include/rclcpp_components/component_manager.hpp                                                    ││                                                                                                                           │
│+++ b/rclcpp_components/include/rclcpp_components/component_manager.hpp                                                    ││                                                                                                                           │
│@@ -99,7 +99,7 @@ public:                                                                                                  ││                                                                                                                           │
│    */                                                                                                                     ││                                                                                                                           │
│   RCLCPP_COMPONENTS_PUBLIC                                                                                                ││                                                                                                                           │
│   ComponentManager(                                                                                                       ││                                                                                                                           │
│-    std::weak_ptr<rclcpp::Executor> executor,                                                                             ││                                                                                                                           │
│+    std::weak_ptr<rclcpp::Executor> executor = nullptr,                                                                   ││                                                                                                                           │
│     std::string node_name = "ComponentManager",                                                                           ││                                                                                                                           │
│     const rclcpp::NodeOptions & node_options = rclcpp::NodeOptions()                                                      ││                                                                                                                           │
│     .start_parameter_services(false)                                                                                      ││                                                                                                                           │
│diff --git a/rclcpp_components/src/component_container_mt.cpp b/rclcpp_components/src/component_container_mt.cpp           ││                                                                                                                           │
│index 8c32c08f..013c0b1c 100644                                                                                            ││                                                                                                                           │
│--- a/rclcpp_components/src/component_container_mt.cpp                                                                     ││                                                                                                                           │
│+++ b/rclcpp_components/src/component_container_mt.cpp                                                                     ││                                                                                                                           │
│@@ -25,8 +25,7 @@ int main(int argc, char * argv[])                                                                        ││                                                                                                                           │
│                                                                                                                           ││                                                                                                                           │
│   auto options = rclcpp::NodeOptions{};         
```

Reply:
> I believe overload(1) default constructs an empty weak_ptr but that isn't a nullptr - just an empty weak_ptr<rclcpp::Executor>.

Ok, let's use that then.
Suggested patch

```
│--- a/rclcpp_components/include/rclcpp_components/component_manager.hpp                                                    ││                                                                                                                           │
│+++ b/rclcpp_components/include/rclcpp_components/component_manager.hpp                                                    ││                                                                                                                           │
│@@ -99,7 +99,7 @@ public:                                                                                                  ││                                                                                                                           │
│    */                                                                                                                     ││                                                                                                                           │
│   RCLCPP_COMPONENTS_PUBLIC                                                                                                ││                                                                                                                           │
│   ComponentManager(                                                                                                       ││                                                                                                                           │
│-    std::weak_ptr<rclcpp::Executor> executor,                                                                             ││                                                                                                                           │
│+    std::weak_ptr<rclcpp::Executor> executor = std::weak_ptr<rclcpp::Executor>(),                                                                   ││                                                                                                                           │
│     std::string node_name = "ComponentManager",                                                                           ││                                                                                                                           │
│     const rclcpp::NodeOptions & node_options = rclcpp::NodeOptions()                                                      ││                                                                                                                           │
│     .start_parameter_services(false)                                                                                      ││                                                                                                                           ```

Reply:
Fixed at https://github.com/ros2/rclcpp/pull/1708/commits/afb94bc57f28d9f0d7c3209dd335ff290e253721